lib/truffle/rubygems/installing_extra_executables.rb

```diff
@@ -0,0 +1,38 @@
+module Gem
+  module InstallingExtraExecutables
+    extend self
+
+    def install_hooks_for(*extra_bin_dirs)
+      install_hook   = -> tool { hook(extra_bin_dirs, tool, :generate_bin) }
+      uninstall_hook = -> tool { hook(extra_bin_dirs, tool, :remove_executables, tool.spec) }
+      Gem.post_install_hooks.push(install_hook)
+      Gem.pre_uninstall_hooks.push(uninstall_hook)
+      [install_hook, uninstall_hook]
+    end
+
+    def remove_hooks(install_hook, uninstall_hook)
+      [Gem.post_install_hooks.delete(install_hook),
+       Gem.pre_uninstall_hooks.delete(uninstall_hook)]
+    end
+
+    private
+
+    def with_bin_dir(obj, new_bin_dir, &block)
+      old_bin_dir = obj.instance_variable_get :@bin_dir
+      obj.instance_variable_set :@bin_dir, new_bin_dir
+      begin
+        block.call
+      ensure
+        obj.instance_variable_set :@bin_dir, old_bin_dir
+      end
```

Comment:
Can `obj` be used concurrently? If so it's potentially unsafe to change `@bin_dir` like this.

Reply:
obj is an installer or uninstaller tool which are used 1 instance per gem, therefore should be fine.